### ISSUE TITLE
[PM-663] add ability to cluster the cleanup jobs

### ISSUE THREAD

**Issue by mayani** [pegasus-isi/pegasus#781]
User request from Yu Huang

> Do you guys know how to fix the clustering size for executable  "cleanup"? I basically want the cleanup job to be always clustered regardless of the parent job's clustering profile.
>
> I want to set it to something like 50 or 100. Because cleanup jobs takes very little time, in the current setting, it depends on the parent job. Some parent job has low clustering profile because each takes a long time to run but cleanup is always like a second. And then turnover is so frequent, it slows down the condor workflow (lots of jobs in POST).
>

**Reporter:** @vahi
**Resolution:** Fixed
**Watchers:**
@vahi
**Attachments:**

**Comment by mayani** [comment]
**Author:** @vahi

I changed the algorithm to create clustered cleanup jobs instead of doing a separate pass of the workflow.

The algorithm now per level of the workflow, creates cleanup jobs as before ( can be at most one per compute/stageout job ) .
However, we now cluster these cleanup jobs when they are added to the workflow.

The user can specify how many cleanup jobs at maximum he wants created  -  max_num_cleanup_jobs_per_level

We use this parameter to calculate how many cleanup jobs are clustered into one clustered cleanup job as follows

number of cleanup jobs in one cluster = ( number of cleanup jobs per level + max_num_cleanup_jobs_per_level -1)/max_num_cleanup_jobs_per_level

For e.g if the algorithm is creating 6 cleanup jobs at a particular level, and the user specifes max_num_cleanup_jobs_per_level as 4, then the number of jobs in each clustered job will be
( 6 + 4 -1 )/4 = 2 . Hence 3 clustered cleanup jobs will be created for that level.

**Comment by mayani** [comment]
**Author:** @vahi

Users can specify pegasus.file.cleanup.clusters.num  to designate the
maximum number of cleanup jobs added per level of the executable
workflow. Defaults to 2.

**Comment by mayani** [comment]
**Author:** @vahi

our nightly builds seem to work fine.

**Comment by mayani** [comment]
**Author:** @vahi

One of our montage test workflows failed yesterday

On debugging it, I noticed that the output of the mAdd job was mosaic_20121030_144115_2961.fits
 deleted by two cleanup nodes

grep mosaic_20121030_144115_2961.fits clean_up_CCG_level_*in

clean_up_CCG_level_10_0.in:gsiftp://obelix.isi.edu/nfs/ccg4/scratch-6-months-purge/bamboo/pegasus/montage/20121030T144123-0700/mosaic_20121030_144115_2961.fits
clean_up_CCG_level_10_1.in:gsiftp://obelix.isi.edu/nfs/ccg4/scratch-6-months-purge/bamboo/pegasus/montage/20121030T144123-0700/mosaic_20121030_144115_2961.fits

Because of this the shrink job mShrink_ID000142 failed as the input file got deleted before the job ran.
The failure was not triggered earlier as in most cases the shrink job ran before any cleanup job was run.

There are two related bugs. the mosaic file should be deleted by only one clustered cleanup job and it should the mShrink job should be a parent of it.

**Comment by mayani** [comment]
**Author:** @vahi

A tar file of the submit directory with the failed workflow

**Comment by mayani** [comment]
**Author:** @vahi

The bug got triggered because the map that tracks what file is deleted by what cleanup node was no longer updated when the original stub cleanup nodes were added. 
We had moved that update to when we created the actual clustered cleanup graph node.

Because of this, the algorithm ended up assigning two cleanup nodes on the same level with the same file to be deleted.

The fix involves checking whether a file is already deleted or not when we create a clustered cleanup node. Also we have to ensure that we don't create cleanup nodes that delete 0 files!
We now return null in that case.
